update-engine/src/engine.rs

```diff
@@ -0,0 +1,779 @@
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2023 Oxide Computer Company
+
+use std::{borrow::Cow, pin::Pin, sync::Mutex, task::Poll};
+
+use debug_ignore::DebugIgnore;
+use derive_where::derive_where;
+use futures::{future::BoxFuture, prelude::*};
+use linear_map::LinearMap;
+use pin_project_lite::pin_project;
+use tokio::{
+    sync::{mpsc, oneshot},
+    time::Instant,
+};
+
+use crate::{
+    errors::ExecutionError,
+    events::{
+        Event, ProgressEvent, ProgressEventKind, StepComponentSummary,
+        StepEvent, StepEventKind, StepInfo, StepInfoWithMetadata, StepOutcome,
+        StepProgress,
+    },
+    AsError, StepSpec,
+};
+
+#[derive_where(Debug)]
+pub struct UpdateEngine<'a, S: StepSpec> {
+    // TODO: for now, this is a sequential series of steps. This can potentially
+    // be a graph in the future.
+    log: slog::Logger,
+    sender: mpsc::Sender<Event<S>>,
+    // This is a mutex to allow borrows to steps to be held by both
+    // ComponentRegistrar and NewStep at the same time. (This could also be a
+    // `RefCell` if a `Send` bound isn't required.)
+    //
+    // There is an alternative way to do this that doesn't use a mutex but
+    // involves no less than three lifetime parameters, which is excessive.
+    steps: Mutex<Steps<'a, S>>,
+}
+
+impl<'a, S: StepSpec> UpdateEngine<'a, S> {
+    /// Creates a new `UpdateEngine`.
+    pub fn new(log: &slog::Logger, sender: mpsc::Sender<Event<S>>) -> Self {
+        Self {
+            log: log.new(slog::o!("component" => "UpdateEngine")),
+            sender,
+            steps: Default::default(),
+        }
+    }
+
+    /// Adds a new step corresponding to the given component.
+    pub fn new_step<F, Fut, T>(
+        &self,
+        component: S::Component,
+        id: S::StepId,
+        description: impl Into<Cow<'static, str>>,
+        step_fn: F,
+    ) -> NewStep<'_, 'a, S, T>
+    where
+        F: FnOnce(StepContext<S>) -> Fut + 'a,
+        Fut: Future<Output = Result<StepResult<S, T>, S::Error>> + Send + 'a,
+        T: Send + 'a,
+    {
+        self.for_component(component).new_step(id, description, step_fn)
+    }
+
+    /// Creates a [`ComponentRegistrar`] that contains the component for a given.
+    pub fn for_component(
+        &self,
+        component: S::Component,
+    ) -> ComponentRegistrar<'_, 'a, S> {
+        ComponentRegistrar { steps: &self.steps, component }
+    }
+
+    /// Executes the list of steps. The sender is a list of steps.
+    pub async fn execute(self) -> Result<(), ExecutionError<S>> {
+        let total_start = Instant::now();
+
+        let steps = {
+            let mut steps_lock = self.steps.lock().unwrap();
+            // Grab the steps and component counts from within steps_lock, then let
+            // steps_lock go. (Without this, clippy warns about steps_lock being
+            // held across await points.)
+            std::mem::take(&mut *steps_lock)
```

Comment:
Yes, there's no concurrency involved here since all the other places borrow the mutex and this owns it. (If we were using `Arc<Mutex<T>>`, this would be a bigger concern -- I specifically wanted to avoid `Arc` for that reason.)

Reply:
Added a comment about this.